### DISCUSSION TITLE
De-duplicate datasource error handling

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#21676] [Suggest an Idea]
Most datasources replicate the same error handling but we could centralize it. Concepts:

- Some hosts are critical/essential automatically (e.g. default registry hosts such as npmjs.org and rubygems.org). Any server-side errors for these (e.g. timeouts, connection lost, 429 or 5xx errors) result in `datasource-failure` errors and aborting the run.
- We should make this concept configurable, e.g. with a new `hostRules` configuration option called `essentialHost`. If set to `true`, it means server errors trigger `datasource-failure` errors. Making it configurable would allow the following improvements compared to today:
  - Users could opt out of triggering datasource-failure due to server errors from a host like npmjs.org or rubygems.org (unlikely)
  - Users could opt in to treating other hosts as essential, e.g. their own Artifactory. They could also add a host rule to make *all* hosts essential (i.e. abort a run due to the server-side failure of any datasource)/
- We should have a default set of error types and codes that trigger server-side errors to be thrown, and allow them to be configurable. Ideally we need a way to replace *or* add codes. This concept would allow users to configure a client error like `404` as a server error instead, or to ignore certain errors we care about by default
- Datasources could remove most try/catch logic and simply pass thrown `got` errors back up to `datasource/index.ts` which can then apply shared logic.

One question is: would these concepts be limited to only *host* rules, or is there also a need for them within *packageRules*? i.e. varying it per-package?

**Comment by rarkins** [comment]
If we named it `abortOnError` instead of `essentialHost`, then it would fit in with both host- and package-based semantics and perhaps be more understandable. A challenge is we don't have a good way to match hosts (registryUrls) using packageRules.

**Comment by rarkins** [comment]
Three possible outcomes from a lookup:
- Success (data)
- Ignored failure (null)
- DatasourceError

Success/data doesn't need any special handling, so we only need to care about thrown errors.

Three types of errors:
- HTTP status codes (500 etc)
- HTTP error codes (ECONNECT etc)
- Error messages (`ParseError` etc)

We could approach it one of two ways:
- Specify which errors to throw DatasourceError on, or
- Specify which errors to ignore

In both cases the default config could be simple, like:
- `ignoreHttpStatusCodes=["*"]`, `ignoreHttpErrorCodes=["*"]`, or
- `throwHttpStatusCodes=[]`, `throwHttpErrorCodes=[]`

Let's say you have a single known registry like `cdnjs` where the only thing you want to ignore is a 404. You could configure it like:

```json
  "cdnjs": {
    "ignoreHttpStatusCodes": ["404"],
    "ignoreHttpErrorCodes": ["ENOTFOUND"]
  }
```

Let's say for some other host that you want to ignore everything except obvious server errors, e.g. 429 and 5xx

```json
{
  "orb": {
    "ignoreHttpStatusCodes": ["400-428", "430-499"]
  }
}
```

Note the above uses strings and a range syntax. Another possibility would be to allow regexes although I'm not sure if any easy way to specify "all 4xx except 429". Supporting ranges may be must simpler as we know they are numbers.

TODO: work out how this configuration approach would work with our existing datasources and hostRules. e.g. you want rules to apply to `registry.npmjs.org` only and not custom registries.

**Comment by rarkins** [comment]
![image](https://user-images.githubusercontent.com/6311784/78456065-f4f91c00-76a1-11ea-847c-1eff562a19f2.png)

**Comment by ghost** [comment]
@rarkins Is there a specific reason for nuget to never throw?

We're running into issues when using expired credentials: setting `abortOnError` to `true` doesn't make Renovate fail because the thrown `ExternalHostError` is pretty much absorbed by v3.ts' `return null` in its `catch`. This makes our Renovate passes fail silently.

Any idea on how we can fix this?

**Comment by viceice** [comment]
@yannik-beaulieu-ia  see #10751

**Comment by ghost** [comment]
@viceice issue is still there. 🙁

Problem happens when Renovate tries to get the ressource URL, [here](https://github.com/renovatebot/renovate/blob/1d2460727374454597693d61ed51e64c36ee390c/lib/datasource/nuget/v3.ts#L94). `getReleases` isn't even called. 😕

**Comment by viceice** [comment]
#10755

**Comment by ghost** [comment]
Renovate now skips properly with code from #10755 but execution still returns 0 as exit code. I suppose this is as designed?

**Comment by viceice** [comment]
@yannik-beaulieu-ia how did you configure `abortOnError` and `abortIgnoreStatusCodes` ?

https://github.com/renovatebot/renovate/blob/c077f15831a1c687aa6132eeb0d9b162f35d5179/lib/util/http/index.ts#L159-L162

**Comment by ghost** [comment]
We're setting host rules via command ligne argument with this value:

``` json
[
  {
    "hostType": "nuget",
    "endpoint": "[url]",
    "username": "[username with expired access to repository]",
    "password": "[password]",
    "abortOnError": true
  }
]
```

`abortOnError` does its job, as we're seeing `INFO: External host error causing abort - skipping` in the logs. Checking the exit code by running `echo %errorlevel%` (after running Renovate) returns 0.

**Comment by viceice** [comment]
I think it's intentional that renovate didn't exit with failed, it just skips the repo for  `ExternalHostError`

**Comment by ghost** [comment]
Yeah, that's why I think it feels as designed.

Thanks for the help!
